Feature/rootless and usability

weeezes opened the pull request:
Hello,

Sorry that this is such a large one, I understand if you'd like me to split it into smaller pieces. This was something I needed so I could make the following job run as I was expecting it to:

```
job "example" {
  datacenters = ["dc1"]
  type = "service"
  
  group "cache" {
    count = 1
    restart {
      attempts = 2
      interval = "30m"
      delay = "15s"
      mode = "fail"
    }
    task "redis" {
      driver = "podman"

      config {
        image = "docker://registry.hub.docker.com/library/redis:3.2"

        port_map {
          db = 6379
        }

        volumes = [
          "alloc/:/mnt/alloc",
          "local/:/mnt/local",
          "secrets/:/mnt/secrets",
        ]
      }

      resources {
        cpu    = 500 # 500 MHz
        memory = 256 # 256MB

        network {
          mbits = 10
          port  "db"  {}
        }
      }
    }
  }
}
```

Which was triggered by my first attempt at running jobs with this driver and comparing the experience to [the intro on podman.io](https://podman.io/):

> What is Podman? Podman is a daemonless container engine for developing, managing, and running OCI Containers on your Linux System. Containers can either be run as root or in rootless mode. Simply put: `alias docker=podman`.

After the changes in this PR I can basically change the driver from `"docker"` to `"podman"` and get the same result, which seemed like a nice idea to me.

At the same time my current podman setup is rootless, so I needed the support for that.

What do you think, is there something here you'd like to have upstream :)?

All tests run for me locally:

```
PASS
ok      github.com/pascomnet/nomad-driver-podman        46.612s
```

But the OOM test is a bit flaky.

towe75 commented:
@weeezes  wow, this is amazing! Thank you for all the work. I will need some time to review and check it.

weeezes commented:
Thank you, another pair of eyes will surely find something I can fix :).

towe75 commented:
I will do a more detailed review but i have already a couple of questions/suggestions:

* you hardcoded uid 1000 for the varlink socket path. Would be better to infer the path from the real uid. Also a fallback to the default root path in case of uid=0 would be nice. 
* swappiness in cgroupv2. Maybe better to keep the parameters for cgroupv1 users and guard the unittest and runtime with a cgroupv2 check.
* i do not understand your change related to allocation local directories. Maybe you can elaborate your intention.

weeezes commented:
> you hardcoded uid 1000 for the varlink socket path. Would be better to infer the path from the real uid. Also a fallback to the default root path in case of uid=0 would be nice.

Very good point, I've clearly gotten a bit tunnel visioned on only getting this to work with my use case :).

> swappiness in cgroupv2. Maybe better to keep the parameters for cgroupv1 users and guard the unittest and runtime with a cgroupv2 check.

I'll figure out a way to do the check, seems like a good thing to check out 👍 . This should probably also have an effect on the varlink socket path, as rootless mode afaik doesn't work with cgroupv1.

> i do not understand your change related to allocation local directories. Maybe you can elaborate your intention.

The intention here was to make it work like it works in the [Docker driver](https://www.nomadproject.io/docs/drivers/docker/#volumes), where relative paths are handled by mapping them to the allocation local directories. Before this change the volumes for me were just `<path_to_nomad_data_dir>/<allocation>/<dir>:<path_to_nomad_data_dir>/<allocation>/<dir>`, which was a bit unexpected. I was expecting `local/conf.json:/mnt/local/conf.json` to lead to `<path_to_nomad_data_dir>/<allocation>/local/conf.json` file being mounted to `/mnt/local/conf.json`.

weeezes commented:
Hi!

I rebased the branch against master so it's now up-to-date with it.

I also addressed the socket path in a way where it now tries to guess what the path might be based on the UID and if we're running on a system supporting cgroup v2 or only v1. Unfortunately I don't have a cgroupv1 system at hand so I've only been able to test it with the included unit tests.

Swappiness is now also set only on cgroupv1 systems, but also I'm not able to test it on my system :sweat_smile: .

weeezes commented on driver.go:
```diff
@@ -327,40 +348,95 @@ func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drive
-		allVolumes = append(allVolumes, driverConfig.Volumes...)
+	d.logger.Debug("Volumes", fmt.Sprintf("%#v", allocVolumes))
+
+	// reset env so we don't inherit host env by default
```
Tested this out and compared it to the Docker driver, and yeah this isn't correct. From my current testing it looks like the right way would be to reset the env to only contain variables prefixed with `NOMAD_*`, then overlay the image defaults on top of those, and lastly pick any task config env variables (which I don't think are implemented yet).

What comes to the volumes I think there's a small difference in how Docker and Podman work here, if we look at how Docker driver handles this: https://github.com/hashicorp/nomad/blob/master/drivers/docker/driver.go#L635-L687 where `docker inspect` shows bind mounts under HostConfig like this:

```
        "HostConfig": {
            "Binds": [
                "/<path-to-nomad-data-dir>/alloc/fd74a661-0766-9c4a-7224-77769b92a0b6/alloc:/alloc:z",
        ...
```

and then separate mounts from that `/alloc` to what's defined in the task config in `volumes = [...]`. It doesn't look like that's needed in Podman, so my reasoning for this was we mount the `/local`, `/secrets` and `/alloc` if it's requested and set the env variables to reflect where they were mounted. Maybe it doesn't make sense this way.

towe75 replied:
I am not sure about the correct overal env handling. It will be best to learn from other drivers. Rawexec, in example, simply calls cfg.EnvList()  and so does the java and qemu drivers. Hence i added the same snipped to the podman driver.

Related to automatic alloc mounts: AFAIK the container should solely rely on env variables to discover Nomads goodies. It would be fine for me mounting it statically on /alloc etc as long as it's correctly reflected in the env.

weeezes replied:
Cool, so that's what I went on and did now after toying around with this for a while :). So the weird env variables I get is actually from running things rootless, it's deriving my users environment which ofc makes it differ from the Docker ones. IMO no need to clear the env now that I realized that's what's happening. The only thing is that the env variables from the image will override the base env variables, which seems to do what's expected for me.

Now the main difference is to handle the task configured volumes like Docker driver does, so e.g. mounting `local/:/conf` gives you a directory within the container where `<nomad-data-dir>/<alloc-local-dir>` is mounted to `/conf` inside the container.

The default alloc mounts are now mounted to `/alloc`, `/local` and `/secrets` and env variables are set to those paths, this is what I noticed Docker driver is doing.

towe75 replied:
It's much better now, thank you!

weeezes commented on driver_test.go:
```diff
@@ -643,26 +672,38 @@ func TestPodmanDriver_OOM(t *testing.T) {
-	// limit memory to 10MB to trigger oom soon enough
-	task.Resources.NomadResources.Memory.MemoryMB = 10
+	// limit memory to 5MB to trigger oom soon enough
+	task.Resources.NomadResources.Memory.MemoryMB = 5
```
Would be ok, I'll keep it lower locally to get it to OOM properly.

weeezes commented on client.go:
```diff
@@ -220,9 +226,72 @@ func (c *PodmanClient) InspectContainer(containerID string) (iopodman.InspectCon
+	}
+
+	cgroupv := 0
+	for _,l := range procFilesystems {
```
Fixed

weeezes commented:
Hey, I had some time to work on the reviewed points today and I think it's in a lot better shape now, thanks for the feedback! If you feel like merging only parts of the PR feel free to go ahead tho :).

towe75 commented on driver.go:
```diff
@@ -327,35 +348,92 @@ func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drive
+	allArgs = append(allArgs, command...)
 	allArgs = append(allArgs, driverConfig.Args...)
+
+	entryPoint := strings.Join(img.Config.Entrypoint, " ")
```
@weeezes  i did a few tests and i think it's not necessary to inspect the image in order to preset entryPoint/workingDir. Seems like podman does "the right thing" if one simply populates createOpts.EntryPoint = nil / createOpts.WorkDir = nil.  Can you please verify this finding on your machine? 
Important is to set it to nil, not to empty string.

weeezes replied:
Nice find, confirmed that that indeed works! Pushed a change for it :).

weeezes commented:
Fixed the `entryPoint` and `workDir` to use the `nil` way to set defaults, seems to work nicely :+1: . Also made the OOM test use the 10MB again so it should work for you too, that passes for me too!

towe75 commented:
Hey @weeezes, nice work!  I will give it a final spin tomorrow. But other people will have to decide about the merge because i'm about to transfer this repo to it's now home over at hashicorp! 

Of course i will continue to collaborate as the original author but other people will hopefully join in very soon.

weeezes commented:
Oh wow, is it going to become official at some point!? That's amazing, great work 🎉🎉! 

Thanks for giving all the great feedback :)!

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/nomad-driver-podman?pullRequest=24)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/nomad-driver-podman?pullRequest=24)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/nomad-driver-podman?pullRequest=24).</sub>

weeezes commented:
Hi, any new reviews needed or something I'd need to do to move this forward :)?

towe75 commented:
@weeezes i can not merge it any longer but i can, of course, continue to review it. 
Few more things:

- command option: I assume that the same nil handling like in the  endpoint and workdir case applies. Can you check? 
- please enhance the documentation, explain the varlink option and also endpoint/workdir.

weeezes commented:
Thanks!


> command option: I assume that the same nil handling like in the endpoint and workdir case applies. Can you check?

It probably would, but the problem is that command is handled as a part of a list of arguments, which is a `[]string`, so it looks like the command has to be set if you want to handle args too. I couldn't at least make this work. If someone finds a way to make it work it should be a quite easy patch for some new contributor to do :).

> please enhance the documentation, explain the varlink option and also endpoint/workdir.

Updated, also did some consistency changes so all the items in the feature list start with the same capitalization. :+1:

towe75 commented:
@weeezes thank you for hanging on. The command stuff still smells a bit, especially because it's the last reason for inspecting the image. My old code simply built a list of arguments for podman, only the first argument is mandatory: the image name. I did not care about the default command or the default entrypoint because podman seems to do the right thing automatically - just like when using the command line. Do you have an actual proof/test that we need to infer the default command from the image? 

Regardind you nix files: there is a hardcoded version number. Would you mind to move the files into a contrib/ folder?

weeezes commented:
> I did not care about the default command or the default entrypoint because podman seems to do the right thing automatically - just like when using the command line. Do you have an actual proof/test that we need to infer the default command from the image?

I need to try this out again, I assumed it would need to have the same "null" way as the other parameters and couldn't make that work. I might have time on the weekend to play around a bit more.

> Regardind you nix files: there is a hardcoded version number. Would you mind to move the files into a contrib/ folder?

Shouldn't be here, I can contribute that at some other point if it's wanted but I've accidentally added it :).

weeezes commented:
Hey, rebased.

About the command:

If you create a container like this:

```
$ ./go-varlink-cmd call unix:/run/user/1000/podman/io.podman/io.podman.CreateContainer '{ "create": { "args": ["docker://registry.hub.docker.com/library/redis:6.0.3-buster"], "detach": true }}'
{
  "container": "2fe17690bf85a46bef341a2d5ba99081aec5e9642e029b12b7138eaf2c4d1fe1"
}

$ ./go-varlink-cmd call unix:/run/user/1000/podman/io.podman/io.podman.StartContainer '{ "name": "2fe17690bf85a46bef341a2d5ba99081aec5e9642e029b12b7138eaf2c4d1fe1" }'
{
  "container": "2fe17690bf85a46bef341a2d5ba99081aec5e9642e029b12b7138eaf2c4d1fe1"
}
```

You'll get

```
            "Cmd": [
                "redis-server"
            ],
```

for a redis container.

If you run

```
$ ./go-varlink-cmd call unix:/run/user/1000/podman/io.podman/io.podman.CreateContainer '{ "create": { "args": ["docker://registry.hub.docker.com/library/redis:6.0.3-buster", "--port", "6379"], "detach": true }}'
{
  "container": "5bf3f442cde191807e36bbed45ca5e25469381e1d0bd04fde05ddb74f4d81296"
}

$ ./go-varlink-cmd call unix:/run/user/1000/podman/io.podman/io.podman.StartContainer '{ "name": "5bf3f442cde191807e36bbed45ca5e25469381e1d0bd04fde05ddb74f4d81296" }'
{
  "container": "5bf3f442cde191807e36bbed45ca5e25469381e1d0bd04fde05ddb74f4d81296"
}

```

you'll get 

```
            "Cmd": [
                "--port",
                "6379"
            ],
```

which is the same as if you just run `podman run -d docker://registry.hub.docker.com/library/redis:6.0.3-buster --port 6379` so looks like what you said is on point and I've had some misunderstanding on how it was handled :).

towe75 commented:
@weeezes good job! 

Regarding your env handling and infering the env from the image. Again i think it's not necessary.
I took the time and ran this test:

a) Create a Dockerfile:

```
FROM ubuntu:18.04
 
ENV test_foo=1
ENV test_bar baz
 
CMD ["bash", "-c", "env"]
```

b) Build/commit it

```podman build -t envtest .```

c) Test on cli

```
git:(master) $ podman run -it --rm envtest 
        
...       
test_foo=1
test_bar=baz                                                                                                                                                                                                                                                                                                          
...
```

d) Create a nomad batch job

```
job "envtest" {
  datacenters = ["dc1"]
  type        = "batch"
  group "envtest" {
      task "envtest" {
      driver = "podman"

      config {
          image = "envtest"
      }
    }
  }
}

```

e) Run it with current master (without your changes), verify output:

```
git:(master) $ nomad run examples/envtest.hcl                                             

...
Allocation "55c6c0c9" created: node "c6c2bf81", group "envtest"                                                                                                             ...

git:(master) $ nomad alloc logs 55c6c0c9|grep test_

2020-06-19T13:27:08.840341873-04:00 stdout F test_foo=1                                                                                                                     2020-06-19T13:27:08.840341873-04:00 stdout F test_bar=baz 
```

As you see: the env from the image is already automatically used.

weeezes commented:
Thanks :). And thanks for your patience!

About the envs: That test indeed works, but I think the problem now is that I am running a rootless environment, so all the containers are inheriting the environment from the parent process. I confirmed this just now, here's the results:

I'm running with this job spec:

```
job "example" {
  datacenters = ["dc1"]
  type = "service"
  
  group "cache" {
    count = 1
    restart {
      attempts = 2
      interval = "30m"
      delay = "15s"
      mode = "fail"
    }
    task "redis" {
      driver = "podman"

      config {
        image = "docker://registry.hub.docker.com/library/redis:6.0.3-buster"

        port_map {
          db = 6379
        }

        volumes = [
          "alloc/:/mnt/alloc",
          "local/:/mnt/local",
          "secrets/:/mnt/secrets",
        ]
      }

      resources {
        cpu    = 500 # 500 MHz
        memory = 256 # 256MB

        network {
          mbits = 10
          port  "db"  {}
        }
      }
    }
  }
}
```

With the change in, the `$PATH` is as follows within the container:

```
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

If I take the change out like this:

```
diff --git a/driver.go b/driver.go
index dba247a..22460b1 100644
--- a/driver.go
+++ b/driver.go
@@ -349,7 +349,7 @@ func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drive
                return nil, nil, fmt.Errorf("image name required")
        }
 
-       img, err := d.createImage(cfg, &driverConfig)
+       _, err := d.createImage(cfg, &driverConfig)
        if err != nil {
                return nil, nil, fmt.Errorf("Couldn't create image: %v", err)
        }
@@ -416,11 +416,11 @@ func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drive
 
        // ensure to include port_map into tasks environment map
        cfg.Env = taskenv.SetPortMapEnvs(cfg.Env, driverConfig.PortMap)
-       // Set the env to image defaults
-       for _,v := range img.Config.Env {
-               p := strings.Split(v, "=")
-               cfg.Env[p[0]] = p[1]
-       }
+       //// Set the env to image defaults
+       //for _,v := range img.Config.Env {
+       //      p := strings.Split(v, "=")
+       //      cfg.Env[p[0]] = p[1]
+       //}
        allEnv := []string{}
        for k,v := range cfg.Env {
                allEnv = append(allEnv, fmt.Sprintf("%s=%s", k, v))
```

and run the same job, first of all the container fails like this:

```
Recent Events:
Time                       Type             Description
2020-06-19T21:28:59+03:00  Killing          Sent interrupt. Waiting 5s before force killing
2020-06-19T21:28:58+03:00  Alloc Unhealthy  Unhealthy because of failed task
2020-06-19T21:28:58+03:00  Not Restarting   Error was unrecoverable
2020-06-19T21:28:58+03:00  Driver Failure   rpc error: code = Unknown desc = failed to start task, could not start container: io.podman.ErrorOccurred(Reason: executable file not found in $PATH: No such file or directory: OCI runtime command not found error)
2020-06-19T21:28:57+03:00  Task Setup       Building Task Directory
2020-06-19T21:28:57+03:00  Received         Task received by client
```

And is clear when I look at the Nomad logs from the driver, `$PATH` is set to my hosts `$PATH`:

```
PATH=.../nix/var/nix/profiles/default/bin:/run/current-system/sw/bin
```

which obviously doesn't work.

When I run the same container with `podman run -d docker://registry.hub.docker.com/library/redis:6.0.3-buster` and inspect the result the `$PATH` is as follows:

```
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "TERM=xterm",
                "REDIS_DOWNLOAD_SHA=bca46dce81fe92f7b2de4cf8ae41fbc4b94fbd5674def7f12c87e7f9165cbb3a",
                "container=podman",
                "GOSU_VERSION=1.12",
                "REDIS_VERSION=6.0.3",
                "REDIS_DOWNLOAD_URL=http://download.redis.io/releases/redis-6.0.3.tar.gz",
                "HOSTNAME=775a026aff61",
                "HOME=/root"
            ],
```

When I run it straight with varlink like this:

```
./go-varlink-cmd call unix:/run/user/1000/podman/io.podman/io.podman.CreateContainer '{ "create": { "args": ["docker://registry.hub.docker.com/library/redis:6.0.3-buster"], "detach": true }}'
./go-varlink-cmd call unix:/run/user/1000/podman/io.podman/io.podman.StartContainer '{ "name": "ac460d44e1a9791f3b1e230b874c620f0a22c3dd789bd126bad28db3a0e776da" }'
```

The `$PATH` is set like it should be when it comes to the `$PATH`, the rest of the env is inherited like it is when I run it through the driver:

```
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

So there clearly is something at play here that requires the order to either be:

0. Optionally clear the env
1. Host env
2. Image env
3. Task config env

I unfortunately can't take that change out unless there's some other way to handle this, because there's no way for me to use this without the env being managed in a working way :). 

I hope this clears this out a bit.

towe75 commented:
@weeezes: i assume that you simply inherit nomads process environment. I'm running it through systemd, so i have a controlled and clean env. Can you maybe compare nomads environment (via /proc/NNN/environ) to see if it is part of the culprit? 

Otherwise i would say that this is a typical situation without a "right or wrong" which usually results in more configuration options.

weeezes commented:
Yes, Nomads system environment. The issue is that I want to keep this tolerant to no matter what the initial state is, because otherwise it does the wrong thing by default. That one four line while loop is crucial to keep it working consistently imo :).

drewbailey commented on driver.go:
```diff
@@ -328,35 +349,83 @@ func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *drive
+	for k,v := range cfg.Env {
+		allEnv = append(allEnv, fmt.Sprintf("%s=%s", k, v))
+	}
+	d.logger.Debug("Env", fmt.Sprintf("%#v", allEnv))
```
Trying to think if there's anything here we wouldn't want logged.

weeezes replied:
Oh yeah, stuff like secrets would get leaked through this. Probably safer to remove it, good catch! Let's see if I can get to doing it in the evening :rocket: .

towe75 commented:
@weeezes: can you please try to apply a env blacklist like documented at https://www.nomadproject.io/docs/configuration/client? 

I agree that it should not be necessary to do this but wiping the entire env is equally wrong.

Maybe the environment handling is different to the docker driver because of the external plugin process. I could not find any cleanup code in the docker driver.

@drewbailey : can you maybe clearify? I tend to file an issue to the nomad repo. Looks like the plugins environment leaks into the task environment.

drewbailey commented:
@towe75 i was only talking about the log statement to prevent printing out sensitive environment variables to the logger. I think having an env allowlist/denylist would be nice eventually but also fine for a follow up issue.

weeezes commented:
> @weeezes: can you please try to apply a env blacklist like documented at https://www.nomadproject.io/docs/configuration/client?

> I agree that it should not be necessary to do this but wiping the entire env is equally wrong.

> Maybe the environment handling is different to the docker driver because of the external plugin process. I could not find any cleanup code in the docker driver.

I think we're going loops here :). It has already been established that it's the rootless environment where I'm running this that doesn't start with a clean environment in our previous conversations.

 Unfortunately I don't have time to try to polish this further if something else is needed than just the removal of the debug log. I couldn't even get to opening my VM in the last few days, life is quite busy currently :). 

Thanks for the comment again @drewbailey , I'll push the change today.

weeezes commented:
Hello, found another useless and potentially harmful debug log which I removed. Thanks for pointing the logging problem out @drewbailey :).

drewbailey commented:
Thanks for the contribution!

drewbailey commented:
included in #42
